File: backend/methods/modified_euler.py
from .utils import evaluate_function, evaluate_system

def modified_euler_method(func_str, x0, y0, h, steps_count):
    """
    Modified Euler Method (Predictor-Corrector)
    """
    steps = []
    x = x0
    y = y0
    
    for i in range(steps_count):
        try:
            f_xy = evaluate_system([func_str], {'x': x, 'y': y})[0]
            # Predictor
            y_predict = y + h * f_xy
            x_next = x + h
            f_xnext_ypredict = evaluate_system([func_str], {'x': x_next, 'y': y_predict})[0]
            
            # Corrector
            y_correct = y + (h / 2) * (f_xy + f_xnext_ypredict)
        except Exception as e:
            return {"success": False, "message": str(e), "steps": steps}
            
        step_data = {
            "iteration": i,
            "x": x,
            "y": y,
            "f_xy": f_xy,
            "y_predict": y_predict,
            "f_xnext_ypredict": f_xnext_ypredict,
            "y_correct": y_correct
        }
        steps.append(step_data)
        
        y = y_correct
        x = x_next
        
    return {
        "success": True,
        "steps": steps,
        "final_x": x,
        "final_y": y
    }

File: worker/src/methods/modifiedEuler.ts
import { evaluateSystem } from "../lib/exprEval";

export interface ModifiedEulerStep {
  iteration: number;
  x: number;
  y: number;
  f_xy: number;
  y_predict: number;
  f_xnext_ypredict: number;
  y_correct: number;
}

export function modifiedEulerMethod(funcStr: string, x0: number, y0: number, h: number, stepsCount: number) {
  const steps: ModifiedEulerStep[] = [];
  let x = x0;
  let y = y0;

  for (let i = 0; i < stepsCount; i++) {
    let fXy: number, yPredict: number, xNext: number, fXnextYpredict: number, yCorrect: number;
    try {
      fXy = evaluateSystem([funcStr], { x, y })[0];
      yPredict = y + h * fXy;
      xNext = x + h;
      fXnextYpredict = evaluateSystem([funcStr], { x: xNext, y: yPredict })[0];
      yCorrect = y + (h / 2) * (fXy + fXnextYpredict);
    } catch (e) {
      return { success: false, message: (e as Error).message, steps };
    }

    steps.push({
      iteration: i,
      x,
      y,
      f_xy: fXy,
      y_predict: yPredict,
      f_xnext_ypredict: fXnextYpredict,
      y_correct: yCorrect,
    });

    y = yCorrect;
    x = xNext;
  }

  return { success: true, steps, final_x: x, final_y: y };
}
